File: pkg/mb64/sort.go
package mb64

import (
	"math"
)

var baseChars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"

func genCharMap(chars string) map[rune]int {
	m := make(map[rune]int)
	for i, c := range chars {
		m[c] = i
	}
	return m
}

func sum(numbers []int) int {
	total := 0
	for _, n := range numbers {
		total += n
	}
	return total
}

func reverse(s string) string {
	runes := []rune(s)
	for i, j := 0, len(runes)-1; i < j; i, j = i+1, j-1 {
		runes[i], runes[j] = runes[j], runes[i]
	}
	return string(runes)
}

var baseCharsMap = genCharMap(baseChars)

func charsToNumbers(chars string) []int {
	numbers := []int{}
	for _, c := range chars {
		if c == '=' {
			numbers = append(numbers, sum(numbers)%64)
		} else {
			numbers = append(numbers, baseCharsMap[c])
		}
	}
	return numbers
}

func sortStr(str string, numbers []int) string {
	res := str
	for _, number := range numbers {
		_res := ""
		_str := res
		for len(_str) > 0 {
			powResult := (number + len(_str)) * int(math.Abs(float64(number-len(_str))))
			index := powResult % len(_str)
			_res += string(_str[index])
			_str = _str[:index] + _str[index+1:]
		}
		_res = reverse(_res)
		res = _res
	}
	return res
}

File: js/src/mb64/sort.ts
const baseChars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

const genCharMap = (chars: string): { [key: string]: number } => {
  const map: { [key: string]: number } = {};
  for (let i = 0; i < chars.length; i++) {
    map[chars[i]] = i;
  }
  return map;
};

const sum = (numbers: number[]): number => {
  return numbers.reduce((acc, curr) => acc + curr, 0);
};

const reverse = (str: string): string => {
  return str.split("").reverse().join("");
};

const baseCharsMap = genCharMap(baseChars);

const charsToNumbers = (chars: string): number[] => {
  const numbers: number[] = [];
  for (const c of chars) {
    if (c === "=") {
      numbers.push(sum(numbers) % 64);
    } else {
      numbers.push(baseCharsMap[c]);
    }
  }
  return numbers;
};

const sort = (str: string, numbers: number[]): string => {
  let res = str;
  for (const number of numbers) {
    let _res = "";
    let _str = res;
    while (_str.length > 0) {
      const p = (number + _str.length) * Math.abs(number - _str.length);
      const index = p % _str.length;
      _res += _str[index];
      _str = _str.slice(0, index) + _str.slice(index + 1);
    }
    _res = reverse(_res);
    res = _res;
  }
  return res;
};

const genNewBaseChars = (password: string): string => {
  const b64 = btoa(password);
  const numbers = charsToNumbers(b64);
  return sort(baseChars, numbers);
};

export { genNewBaseChars };
